### ISSUE TITLE
Should Dependabot be updating lock (such as Gemfile.lock) files?

### ISSUE THREAD

**Issue by ThomasOwens** [dependabot/dependabot-core#7572]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

Bundler

### Manifest location and content before the Dependabot update

The `/Gemfile.lock` file was updated for a new version of a dependency.

```
diff --git a/Gemfile.lock b/Gemfile.lock
index 56f03bb..a407842 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -195,7 +195,7 @@ GEM
       rspec-mocks (~> 3.12)
       rspec-support (~> 3.12)
     rspec-support (3.12.1)
-    rubocop (1.54.1)
+    rubocop (1.54.2)
       json (~> 2.3)
       language_server-protocol (>= 3.17.0)
       parallel (~> 1.10)
```

### dependabot.yml content

```
version: 2
updates:
  - package-ecosystem: "bundler"
    directory: "/"
    schedule:
      interval: "weekly"
```

### Updated dependency

rubocop from 1.54.1 to 1.54.2

### What you expected to see, versus what you actually saw

It's not clear to me if Dependabot updates transitive dependencies.

What I saw when running `bundler outdated --strict` and `bundler update` for rubocop matches, but these commands also updated other dependencies. However, it only shows me that this particular case works. It does not demonstrate that all necessary rules for transitive dependencies would be applied or rules across the Gemfile would be applied.

At this point, I don't have any evidence that I can trust Dependabot's PRs to do all of the necessary updates. At most, I can use them as a flag to run `bundler outdated --strict` and `bundler update` and apply those changes. And at this point, I'd rather have an issues created and managed rather than a PR.

### Native package manager behavior

I ran `bundle outdated --strict` on the same branch. This showed 4 gems that could be updated: irb, public_suffix, rspec-mocks, and rubocop. Of these, only rubocop is in the Gemfile. The others are transitive dependencies that can be updated and still meet the requirements in the Gemfile.

The output for rubocop was exactly what Dependabot suggested. This only indicates that no transitive updates needed to be updated in a way that changed the ruled in the Gemfile.lock and says nothing about the correctness of the behavior should they need to be updated.

**Comment by armin-joellenbeck** [comment]
Have you tried the following config in dependabot.yml:

```code
    allow:
      - dependency-type: all
```

**Comment by ThomasOwens** [comment]
@armin-joellenbeck Looking at the documentation, I don't see how that helps.

Dependabot shouldn't be touching lock files. For a Ruby project, Dependabot should only be looking at the Gemfile. Similarly, for a Node project, it should only be looking at the package.json file. Although it is good to know if an update exists in a transitive dependency, it is dangerous to update a transitive dependency by editing a lock file. The packages that use that dependency may not work with a different version. You could make a claim that if everyone is using semantic versioning and has correct version pinning, it "should work", but that's a risky assumption.

That doesn't mean that Dependabot shouldn't look at lock files. Knowing that a transitive dependency is outdated or has a vulnerability is a good thing. But if an update is going to be suggested, it needs to be updated in a safe way, and making changes to a lock file is not safe. The default behavior should not be an unsafe operation.

If I'm mistaken, I'd like someone to point me to why updating a lock file is a safe and accepted practice. Otherwise, I think the right course of action is to ensure that Dependabot pull requests are only made against the user-controlled dependency file. If desired, issues could be open to inform the author of outdated or vulnerable transitive dependencies so the maintainer can review and take the appropriate action.

**Comment by armin-joellenbeck** [comment]
I think the proposed config makes the difference.

With it dependabot updates also indirect dependencies in Gemfile.lock.

Without it dependabot only updates dependencies in Gemfile.

Which choice is preferable, is up to you. I tend to the first one. But there are valid reasons to prefer the second.

I hope I have not misunderstood your pupose. Sorry, if this is the case.

**Comment by ThomasOwens** [comment]
@armin-joellenbeck OK, I think I see what you're saying.

If I'm understanding, you're suggesting that Dependabot may be acting like [`bundle update --conservative`](https://bundler.io/v2.5/man/bundle-update.1.html) by default. This prevents Bundler from updating indirect/transitive dependencies.

If I make your proposed change, then Dependabot should act like `bundle update`.

Coming from a Ruby background, this is very strange behavior. I can't remember a time when I ran `bundle update --conservative` - I don't think I ever had. It's not part of [the recommended workflow](https://bundler.io/v2.5/man/bundle-update.1.html#RECOMMENDED-WORKFLOW), either.

I do have to experiment to confirm, but I'd still consider this a bug in Dependabot. It doesn't fit what I would consider to be the expectations of a Ruby developer. For the bundler ecosystem, if you are correct, then `dependency-type: all` should be the default behavior, since that is closer to the standard workflow for a Ruby developer updating dependencies on their own. It would be extremely helpful if the default, out-of-the-box behavior of Dependabot mirrors the manual behaviors of Ruby developers.

**Comment by armin-joellenbeck** [comment]
I see, that we agree about the preferred workflow. But if the developers of dependabot prefer the conservative approach as default, it is ok for me, because it is configurable.

Nonetheless I am interested in the outcome of your experiments.

**Comment by dentarg** [comment]
@jeffwidman @jurre is it possible to configure Dependabot to run `bundler` with [`--conservative`](https://bundler.io/v2.7/man/bundle-update.1.html)?

> Use bundle install conservative update behavior and do not allow indirect dependencies to be updated.

<details><summary>Example: <code>webmock</code> bump in my project (<a href="https://github.com/Starkast/wikimum/pull/817">PR</a>)</summary>
<p>

```shell
$ bundle update webmock
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Resolving dependencies...
Fetching bigdecimal 3.3.1 (was 3.2.2)
Fetching public_suffix 6.0.2 (was 6.0.1)
Fetching hashdiff 1.2.1 (was 1.1.2)
Installing bigdecimal 3.3.1 (was 3.2.2) with native extensions
Installing public_suffix 6.0.2 (was 6.0.1)
Installing hashdiff 1.2.1 (was 1.1.2)
Fetching crack 1.0.1 (was 1.0.0)
Installing crack 1.0.1 (was 1.0.0)
Fetching webmock 3.26.0 (was 3.25.1)
Installing webmock 3.26.0 (was 3.25.1)
Updating files in vendor/cache
  * webmock-3.26.0.gem
  * bigdecimal-3.3.1.gem
  * public_suffix-6.0.2.gem
  * crack-1.0.1.gem
  * hashdiff-1.2.1.gem
Removing outdated .gem files from vendor/cache
  * bigdecimal-3.2.2.gem
  * crack-1.0.0.gem
  * hashdiff-1.1.2.gem
  * public_suffix-6.0.1.gem
  * webmock-3.25.1.gem
Bundle updated!
1 installed gem you directly depend on is looking for funding.
  Run `bundle fund` for details
```

```shell
$ bundle update --conservative webmock
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Resolving dependencies...
Using webmock 3.26.0 (was 3.25.1)
Updating files in vendor/cache
  * webmock-3.26.0.gem
Removing outdated .gem files from vendor/cache
  * webmock-3.25.1.gem
Bundle updated!
1 installed gem you directly depend on is looking for funding.
  Run `bundle fund` for details
```

</p>
</details>

**Comment by dentarg** [comment]
> [@jeffwidman](https://github.com/jeffwidman) [@jurre](https://github.com/jurre) is it possible to configure Dependabot to run `bundler` with [`--conservative`](https://bundler.io/v2.7/man/bundle-update.1.html)?

I don't think it is possible. I created a feature request: https://github.com/dependabot/dependabot-core/issues/13664
